obsplus/structures/grid.py

```diff
@@ -0,0 +1,1323 @@
+    translate_x : float
+        Amount to move the coordinates in the x-direction\n
+    translate_y : float
+        Amount to move the coordinates in the y-direction\n
+    translate_z : float
+        Amount to move the coordinates in the z-direction\n
+    rotate_xy : float
+        Angle (in radians) to rotate the coordinates in the xy plane.
+        Positive is CCW.\n
+    rotate_xz : float
+        Angle (in radians) to rotate the coordinates in the xz plane.
+        Positive is CCW.\n
+    rotate_yz : float
+        Angle (in radians) to rotate the coordinates in the yz plane.
+        Positive is CCW.\n
+    project : dict
+        Strings for the pyproj.Proj class projection (ex:
+        "+init=EPSG:32611"). Keys in the dictionary should be "from" and
+        "to"\n
+
+    Returns
+    -------
+    pandas Dataframe :
+        Dataframe (either a copy or the original) with the converted coordinates
+    """
+    valid_keys = [
+        "scale_x",
+        "scale_y",
+        "scale_z",
+        "translate_x",
+        "translate_y",
+        "translate_z",
+        "rotate_xy",
+        "rotate_yz",
+        "rotate_xz",
+        "project",
+    ]
+    # Copy the input if not inplace
+    if not inplace:
+        points = copy(points)
+
+    # Define the columns to do the conversion on
+    points[xout] = points[xcol]
+    points[yout] = points[ycol]
+    points[zout] = points[zcol]
+
+    # Deal with a callable conversion
+    if isinstance(conversion, Callable):
+        try:
+            if conversion_kwargs is not None:
+                points = conversion(points, **conversion_kwargs)
+            else:
+                points = conversion(points)
+        except Exception as e:
+            raise RuntimeError(f"Conversion callable raised an exception: {e}")
+        if not isinstance(points, pd.DataFrame):
+            raise ValueError("Callable conversion must return a pandas DataFrame")
+        return points
+
+    # Do some checking to make sure the conversion is kosher
+    if not isinstance(conversion, Iterable):
+        raise TypeError(f"conversion shoud be a list of transformations")
+    try:
+        if not conversion[0][0].lower() in valid_keys:
+            raise TypeError(f"conversion should be a list of transformations")
+    except AttributeError:
+        raise TypeError(f"conversion should be a list of transformations")
+
+    # Go through each stage of the conversion
+    for stage in conversion:
+        # Make sure the stage is valid
+        key = stage[0].lower()
+        value = stage[1]
+        if key not in valid_keys:
+            raise ValueError(f"Invalid conversion operation: {key}")
+        elif not key == "project":
+            try:
+                value = float(value)
+            except ValueError:
+                raise TypeError(f"value must be a float for key: {key}, {value}")
+        # Do the conversion
+        if key == "scale_x":
```

Comment:
Can they be replaced by a dict somehow?

Reply:
Are you referring to the conversion? Maybe, but we would need to be pretty careful because the order is extremely important.